### PR TITLE
Added Depth First Search in Typescript

### PR DESCRIPTION
**Pull request by FaMoSi**
**Hi!**  
Thanks for considering for contributing to this ever-growing list of algorithm and data structure implementations.
Your contribution is valuable. Use the follow template for creating PRs. (Delete this part)
---

**Fixes #{ISSUE_NUMBER}**

By submitting this pull request I confirm I've read and complied with the below declarations.

-   [x] I have read the [Contribution guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md) 
and I am confident that my PR reflects them.

-   [x] I have followed the [coding guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#cs) 
for this project.

-   [x] My code follows the [skeleton code structure](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#sample).

-   [x] This pull request has a descriptive title. 
For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.

-   [x] This pull request shall only be merged once the build passes.

-   [x] I have mentioned the issue number correctly (with hyper-link) in this pull request description.

### CODE REVIEW

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        this.adjacencies[node1Label].push(node2Label);
+    }
+  
+    /**
+     * Depth first search
+     *
+     * @param {Function} [cb]
+     * @returns {array}
+     */
+    depthFirstSearch(cb?: Function){
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:72]: Codacy found an issue: [Method 'depthFirstSearch' has a complexity of 6. Maximum allowed is 5.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    }
+  
+    /**
+     * Set a connection between two nodes
+     *
+     * @param {number} node1
+     * @param {number} node2
+     */
+    setConnection(node1Label: number, node2Label: number){
+        if(!this.allowCyclic && node1Label === node2Label) throw Error('Graph cannot be cyclic');
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:62]: Codacy found an issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    
+            // process the node if there is a callback
+            if(cb){
+            const result = cb(node, visitedNodes, nodesToProcess);
+            if(result === true) break;
+            }
+    
+            if(!(visitedNodes.indexOf(node) > -1)){
+            visitedNodes.push(node);
+            if(!this.adjacencies[node]) continue;
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:86]: Codacy found an issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  
+    /**
+     * Depth first search
+     *
+     * @param {Function} [cb]
+     * @returns {array}
+     */
+    depthFirstSearch(cb?: Function){
+        const visitedNodes : Array<number> = [];
+        let nodesToProcess : Array<number> = [0]
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:74]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        while(nodesToProcess.length){
+            let node = nodesToProcess.shift();
+    
+            // process the node if there is a callback
+            if(cb){
+            const result = cb(node, visitedNodes, nodesToProcess);
+            if(result === true) break;
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:81]: Codacy found an issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    tree.setConnection(0, 2);
+    tree.setConnection(2, 6);
+    tree.setConnection(1, 3);
+    tree.setConnection(1, 4);
+    tree.setConnection(1, 5);
+    tree.setConnection(3, 7);
+    tree.setConnection(4, 8);
+    tree.setConnection(4, 9);
+
+    console.log("Visited Nodes DFS:\n", JSON.stringify(tree.depthFirstSearch()));
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:125]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts (hunk 2 above)
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:62]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+main()
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:128]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        this.adjacencies[node1Label].push(node2Label);
+    }
+  
+    /**
+     * Depth first search
+     *
+     * @param {Function} [cb]
+     * @returns {array}
+     */
+    depthFirstSearch(cb?: Function){
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:72]: Codacy found an issue: [Method 'depthFirstSearch' has a complexity of 6. Maximum allowed is 5.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    }
+  
+    /**
+     * Set a connection between two nodes
+     *
+     * @param {number} node1
+     * @param {number} node2
+     */
+    setConnection(node1Label: number, node2Label: number){
+        if(!this.allowCyclic && node1Label === node2Label) {throw Error("Graph cannot be cyclic")};
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:62]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    tree.setConnection(1, 4);
+    tree.setConnection(1, 5);
+    tree.setConnection(3, 7);
+    tree.setConnection(4, 8);
+    tree.setConnection(4, 9);
+
+    console.log("Visited Nodes DFS:\n", JSON.stringify(tree.depthFirstSearch()));
+}
+
+main()
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:128]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    
+            // process the node if there is a callback
+            if(cb){
+            const result = cb(node, visitedNodes, nodesToProcess);
+            if(result === true) {break};
+            }
+    
+            if(!(visitedNodes.indexOf(node) > -1)){
+            visitedNodes.push(node);
+            if(!this.adjacencies[node]) {continue};
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:86]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+  
+    /**
+     * Depth first search
+     *
+     * @param {Function} [cb]
+     * @returns {array}
+     */
+    depthFirstSearch(cb?: Function){
+        const visitedNodes : Array<number> = [];
+        let nodesToProcess : Array<number> = [0]
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:74]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts (hunk 11 above)
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:86]: Codacy found an issue: [Unnecessary semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+        while(nodesToProcess.length){
+            let node = nodesToProcess.shift();
+    
+            // process the node if there is a callback
+            if(cb){
+            const result = cb(node, visitedNodes, nodesToProcess);
+            if(result === true) {break};
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:81]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:81]: Codacy found an issue: [Unnecessary semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts (hunk 9 above)
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:62]: Codacy found an issue: [Unnecessary semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    tree.setConnection(0, 2);
+    tree.setConnection(2, 6);
+    tree.setConnection(1, 3);
+    tree.setConnection(1, 4);
+    tree.setConnection(1, 5);
+    tree.setConnection(3, 7);
+    tree.setConnection(4, 8);
+    tree.setConnection(4, 9);
+
+    console.log("Visited Nodes DFS:\n", JSON.stringify(tree.depthFirstSearch()));
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:125]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+  tree.setConnection(2, 6);
+  tree.setConnection(1, 3);
+  tree.setConnection(1, 4);
+  tree.setConnection(1, 5);
+  tree.setConnection(3, 7);
+  tree.setConnection(4, 8);
+  tree.setConnection(4, 9);
+}
+
+main()
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:126]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+  
+    /**
+     * Depth first search
+     *
+     * @param {Function} [cb]
+     * @returns {array}
+     */
+    depthFirstSearch(cb?: Function){
+      const visitedNodes : Array<number> = [];
+      let nodesToProcess : Array<number> = [0]
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:74]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+
+        // process the node if there is a callback
+        if(cb){
+          const result = cb(node, visitedNodes, nodesToProcess);
+          if(result === true) {break};
+        }
+
+        if(!(visitedNodes.indexOf(node) > -1)){
+          visitedNodes.push(node);
+          if(!this.adjacencies[node]) {continue};
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:86]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+    }
+  
+    /**
+     * Set a connection between two nodes
+     *
+     * @param {number} node1
+     * @param {number} node2
+     */
+    setConnection(node1Label: number, node2Label: number){
+      if(!this.allowCyclic && node1Label === node2Label) {throw Error("Graph cannot be cyclic")};
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:62]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts (hunk 17 above)
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:86]: Codacy found an issue: [Unnecessary semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+    depthFirstSearch(cb?: Function){
+      const visitedNodes : Array<number> = [];
+      let nodesToProcess : Array<number> = [0]
+      while(nodesToProcess.length){
+        let node = nodesToProcess.shift();
+
+        // process the node if there is a callback
+        if(cb){
+          const result = cb(node, visitedNodes, nodesToProcess);
+          if(result === true) {break};
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:81]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:81]: Codacy found an issue: [Unnecessary semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+      this.adjacencies[node1Label].push(node2Label);
+    }
+  
+    /**
+     * Depth first search
+     *
+     * @param {Function} [cb]
+     * @returns {array}
+     */
+    depthFirstSearch(cb?: Function){
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:72]: Codacy found an issue: [Method 'depthFirstSearch' has a complexity of 6. Maximum allowed is 5.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts (hunk 18 above)
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:62]: Codacy found an issue: [Unnecessary semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:74]: Codacy found an issue: [Method 'depthFirstSearch' has a complexity of 6. Maximum allowed is 5.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

File: algorithms/gr-dfsrh/typescript/depth_first_search.ts [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/typescript/depth_first_search.ts:74]: Codacy found an issue: [Method 'depthFirstSearch' has a complexity of 6. Maximum allowed is 5.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6329384)

**Review by RavinduSachintha — APPROVED** [review]

**Comment by RavinduSachintha** [comment]
@FaMoSi would you be able to add this contribution to the algorithm chart in the main readme file (don't change anything else). then we can merge this PR. :+1:

**Comment by FaMoSi** [comment]
> @FaMoSi would you be able to add this contribution to the algorithm chart in the main readme file (don't change anything else)

Done! :)

**Comment by Danushka96** [comment]
@all-contributors please add @FaMoSi for code and doc

**Comment by Danushka96** [comment]
Hi @FaMoSi 
Thanks for the contribution, Keep it up!

**Pull request merged**